ct: Reader and batch cache

Lazin opened the pull request:
This pull request introduces a new batch cache component within the cloud topics subsystem, significantly improving data retrieval efficiency.

The new component leverages the existing batch_cache and establishes a dedicated batch_cache_index instance for each Cloud Topic Partition (CTP). While these indexes are currently persistent for simplicity, future enhancements will include eviction policies. Importantly, cached record batches within this system can be evicted, maintaining consistency with existing partition behavior.

Furthermore, this PR introduces cloud_topic_partition_reader_impl, a new record batch reader type. This reader efficiently fetches and caches extents from the underlying partition, with the ability to refetch as needed. It progressively materializes record batches stored in L0 objects (with current support limited to L0). Crucially, it can also utilize the new record batch cache, ensuring optimized fetching of recent data for tailing consumers.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
* none

Lazin committed:
storage: Make create_cache method public

The `create_cache` method of `log_manager` is used to create
per-segment `batch_cache_index` instances. It was originally private,
as it served internal storage subsystem needs.

In the `cloud-topics` project, we also need to populate the record
batch cache. This requires access to `log_manager`'s cache creation
logic to ensure consistency with the existing storage behavior.

Creating a separate `batch_cache` for cloud-topics would complicate
memory reclamation. Multiple independent caches would require a
coordinated reclaimer to balance memory across systems, introducing
unnecessary complexity.

By making `create_cache` public, we enable reuse of the existing
cache infrastructure while keeping memory management centralized and
predictable.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Add cloud_topics::batch_cache

This commit introduces the `batch_cache` component to the
cloud_topics subsystem. It serves as a lightweight wrapper around
`storage::batch_cache_index`, creating one index instance per
partition.

Unlike the storage layer, which creates a `batch_cache_index` per
segment, cloud_topics uses a single index per partition. This design
simplifies the initial implementation by ensuring a one-to-one mapping
between partitions and their caches, with no eviction or rebalancing
logic.

Currently, if a partition moves to another shard, its associated
`batch_cache_index` remains in the map. This behavior will be refined
in future commits as part of improved memory management and cleanup.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Create batch_cache object in application

This commit adds support for constructing the `batch_cache` object in
the cloud_topics application. To do this, a new parameter,
`storage::api`, is propagated through the constructor, allowing access
to the `storage::log_manager`.

Additionally, two new methods are added to the
`cloud_topics::data_plane_api` interface: `cache_get` and `cache_put`.
These are thin wrappers that forward calls to the underlying
`batch_cache`, serving as basic plumbing to expose cache access
through the public API.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
k/d: Add cloud_topic_partition_reader

This commit adds reader implementation. The reader fetches the metadata
from the underlying metadata storage (currently only from the underlying
partition). Then it materializes the batches gradually and returns them
to the client.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
k/d: Use cloud_topic_partition_reader

Use reader in the cloud_topic_partition::make_reader implementation.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
k/d: Use record batch cache in cloud topics

Use cache in cloud_topic_partition and
cloud_topic_partition_reader_impl. The cache is only used if it's not
disabled for the NTP or for the broker. The reacer could be configured
to skip the cache entirely.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Demote log messages from error to warn

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

rockwotj commented on src/v/cloud_topics/batch_cache/batch_cache.h:
```diff
@@ -0,0 +1,54 @@
+    // per segment). Here we only have one index per cloud storage partition.
+    // From what I see it should be OK to use index this way. Likely even more
+    // efficient compared to index per segment.
+    absl::btree_map<model::ntp, storage::batch_cache_index_ptr> _index;
```
why not chunked hash map?

rockwotj replied:
I see the note about later commits to cleanup this index, is that coming in another PR?

Lazin replied:
I added something to this PR

Lazin replied:
changed to chunked map

rockwotj replied:
Thanks!

rockwotj commented on src/v/cloud_topics/data_plane_api.h:
```diff
@@ -50,6 +50,13 @@ class data_plane_api {
       std::chrono::milliseconds timeout)
       = 0;
+
+    /// Cache materialized record batch
+    virtual void cache_put(const model::ntp&, const model::record_batch& b) = 0;
+
+    /// Retrieve materialized record batch from cache
+    virtual std::optional<model::record_batch>
+    cache_get(const model::ntp&, model::offset o) = 0;
```
I haven't seen where this is going to be used yet, but feels like it should be internal to `write_and_debounce` and `materialize` right?

Lazin replied:
`write_and_debounce` deals with unordered data which doesn't have kafka offset but our record-batch-cache stores data keyed by offset.  Originally, the idea was to put it there but it complicates things too much.

rockwotj replied:
Makes sense

rockwotj commented on src/v/kafka/data/cloud_topic_partition_reader.cc:
```diff
@@ -0,0 +1,316 @@
+
+namespace kafka {
+
+/*
```
Should this comment go in the header?

Lazin replied:
done

rockwotj commented on src/v/kafka/data/cloud_topic_partition_reader.cc:
```diff
@@ -0,0 +1,316 @@
+          std::current_exception());
+        co_return state::end_of_stream_state;
+    }
+    co_return _meta.empty() ? state::end_of_stream_state : state::ready_state;
```
I find it strange that some internal state is mutated in this method and other state is returned to be mutated in the parent. Why not just make this method self contained?

Lazin replied:
fixed

rockwotj commented on src/v/kafka/data/cloud_topic_partition_reader.cc:
```diff
@@ -0,0 +1,316 @@
+        size_t materialized_bytes = 0;
+        chunked_vector<experimental::cloud_topics::extent_meta> to_materialize;
+        chunked_vector<model::record_batch_header> to_materialize_headers;
+        while (materialized_bytes < _config.max_bytes && !_meta.empty()) {
```
shouldn't `materialized_bytes` be a member because if we materialize many batches (ie. do load slice is called many times) then we want to go to EOF after we've hit the limit right?

Lazin replied:
The `_config.max_bytes` and `_config.bytes_consumed` are used for that. If do load slice is called many times it will just go through the state transitions of the STM which will prevent it from invoking this method if some data is already materialized and stored in the `_batches` field.

rockwotj commented on src/v/kafka/data/cloud_topic_partition_reader.cc:
```diff
@@ -0,0 +1,316 @@
+        // Patch materialized record batches
+        auto batches = std::move(mat_res.value());
+        for (size_t i = 0; i < batches.size(); i++) {
+            auto& data_hdr = batches.at(i).header();
+            auto size = data_hdr.size_bytes;
+            auto crc = data_hdr.crc;
+            data_hdr = to_materialize_headers.at(i);
+            data_hdr.type = model::record_batch_type::raft_data;
+            data_hdr.size_bytes = size;
+            data_hdr.crc = crc;
+            // Recalculate the header crc
+            data_hdr.header_crc = model::internal_header_only_crc(data_hdr);
```
Is this the second place we do this? Should we be pulling these out into standalone helper methods to go placeholder->batch and batch->placeholder instead of having a couple copies?

Then we could have unit tests for the conversion logic as well

Lazin replied:
the code is moved from the other place here

Lazin replied:
It's probably makes sense to introduce some utility that populates record batch headers and converts things. But I'd prefer to add it in a separate PR.

rockwotj replied:
Another PR is OK to me. I feel like it will make it easier to read what is going on without these details mixed in.

rockwotj commented on src/v/kafka/data/cloud_topic_partition_reader.cc:
```diff
@@ -0,0 +1,316 @@
+        co_return state::end_of_stream_state;
+    }
+
+    co_return state::materialized_state;
```
same question as above, I don't get the split of where members are mutated.

Lazin replied:
fixed

rockwotj commented on src/v/kafka/data/cloud_topic_partition.cc:
```diff
@@ -361,16 +413,34 @@ static ss::future<> bg_upload_and_replicate(
+                // We know that the data is replicated so it's safe to add
+                // the batch to the record batch cache before returning.
+                if (cache_enabled) {
+                    // NOTE: the assumption is that the cached term matches
+                    // the actual term. If this is not the case the replication
+                    // should fail.
```
hrm are there races possible here? cc: @mmaslankaprv and @bharathv 

Maybe we should add the term to `cluster::kafka_result`

mmaslankaprv replied:
The code in its current form is incorrect, the term that is snapshotted at the time of lambda capture may not be the one data were written with. There is a possibility to return `term` from Raft replicate result or you can use replicate with term but then the replicate will always fail when the term changes between the time it was called and it returned the result. Even if using the replicate with term, the term must be snapshotted before calling replicate. 
So there are two options:
- return term from Raft
- use replicate with term
I am not fan of those APIs as this is a leak of abstraction but we know have term everywhere so i guess returning it in `replicate` result doesn't sound bad

Lazin replied:
I added `last_term` field to the `replicate_result` and to the `kafka_result`. The code in the PR now uses this field propagated to the caller after replication succeeds.

rockwotj commented on src/v/kafka/data/cloud_topic_partition_reader.cc:
```diff
@@ -74,11 +76,27 @@ cloud_topic_partition_reader_impl::cloud_topic_partition_reader_impl(
   , _underlying(std::move(underlying))
-  , _ct_api(std::move(ct_api)) {}
+  , _ct_api(std::move(ct_api)) {
+    if (_config.max_bytes == 0) {
+        _config.max_bytes = L0_default_bytes_per_fetch;
+    }
+    _config.max_bytes = std::clamp(
+      _config.max_bytes, 0UL, L0_max_bytes_per_fetch);
+}
```
what's up with this? It's not mentioned why this happened in the comit

Lazin replied:
that's just a sanity check for the max_bytes, I'll add a comment

rockwotj replied:
I think max_bytes comes from the client, is there a clamp like this in other places? I am not sure if the kafka server does this or if there is something in local storage for this?

Lazin replied:
I reworked this to be in line with the reset of the code.

rockwotj commented on src/v/kafka/data/cloud_topic_partition_reader.cc:
```diff
@@ -109,6 +127,61 @@ cloud_topic_partition_reader_impl::do_load_slice(
+        }
+        if (
+          !ret.empty() && batch_size + materialized_bytes > _config.max_bytes) {
+            // The batch will cause overflow. We want to accept the oversized
```
`s/overflow/over the limit`?

Lazin replied:
done

Lazin committed:
ct: Evict empty batch_cache indexes periodically

Each NTP in the `batch_cache` maintains a corresponding `batch_cache_index`.
Over time, these index instances may become empty due to asynchronous memory
reclamation. Retaining such empty entries unnecessarily consumes memory by
keeping both the object and its key in the map.

This commit introduces a periodic cleanup mechanism to evict empty
`batch_cache_index` instances from the `batch_cache`. The cleanup is triggered
infrequently and designed to avoid impacting active partitions.

Typically, a large number of `batch_cache_index` instances are created per
shard—one per segment. However, for cloud topics, the index is created per
partition, resulting in significantly lower memory pressure. Empty indexes are
most likely to occur when leadership changes or when partitions become inactive
for extended periods.

Active partitions should remain unaffected by this change unless memory
pressure is high and the batch cache is being actively reclaimed.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/68674-->
<!--c0912f14-3211-46cf-9d44-f993eb3440ff-->
<!--title-->
## Retry command for [Build#68674](https://buildkite.com/redpanda/redpanda/builds/68674)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cloud_topics/e2e_test.py::EndToEndCloudTopicsTest.test_write
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--80f394bc-a27a-4a01-85f9-4e91a759d416-->
<!--TEST RESULTS-->
<!--33be727a-2166-4e5e-a9b2-2d9718103a43-->
<!--TEST RESULTS-->
<!--f94aec07-0247-4853-a4d7-4257ef1e8d14-->
<!--TEST RESULTS-->
<!--6d900279-2712-4811-9a29-e560ba6bc9fc-->
<!--TEST RESULTS-->
<!--3bc274a3-13b6-4a72-a384-079eb61fb88e-->
<!--TEST RESULTS-->
<!--61bf2eb7-f667-4460-b31c-5e9a588c3133-->
<!--TEST RESULTS-->
<!--4d137cc8-026b-4d5e-a50a-4c0615c13f0e-->
<!--TEST RESULTS-->
<!--6858a6d0-5efe-464c-b0f0-794a728ba792-->
<!--TEST RESULTS-->
<!--09d2d2fa-5c53-42cb-a69f-1b9745618193-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68674">build#68674</a></summary>


| test_class              | test_method   | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                      |
|:------------------------|:--------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTest | test_write    |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68674#0197f10c-56ed-4b84-b604-b5c40b396c14 | FLAKY         | 6/21     | upstream reliability is '100.0'. current run reliability is '28.57142857142857'. drift is 71.42857 and the allowed drift is set to 50. The test should FAIL |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68684">build#68684</a></summary>


| test_class                  | test_method                                | test_arguments                                                      | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------------------|:-------------------------------------------|:--------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTest     | test_write                                 |                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68684#0197f192-d125-4cc3-87d9-a4ce8f4c1c5c | FLAKY         | 3/21     | upstream reliability is '100.0'. current run reliability is '14.285714285714285'. drift is 85.71429 and the allowed drift is set to 50. The test should FAIL           |
| DeleteRecordsTest           | test_delete_records_concurrent_truncations | {"cloud_storage_enabled": false, "truncate_point": "random_offset"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68684#0197f192-d124-43d6-99b8-d7a4d6dc1112 | FLAKY         | 20/21    | upstream reliability is '99.70930232558139'. current run reliability is '95.23809523809523'. drift is 4.47121 and the allowed drift is set to 50. The test should PASS |
| TopicDeleteCloudStorageTest | drop_lifecycle_marker_test                 | {"cloud_storage_type": 2}                                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68684#0197f192-d12a-4d83-8771-44c331065435 | FLAKY         | 20/21    | upstream reliability is '97.9253112033195'. current run reliability is '95.23809523809523'. drift is 2.68722 and the allowed drift is set to 50. The test should PASS  |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68730">build#68730</a></summary>


| test_class               | test_method             | test_arguments                                                                                                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                     |
|:-------------------------|:------------------------|:---------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTest  | test_write              |                                                                                                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68730#0197f4a2-fbbc-4d4e-b91c-a568e426e734 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                     |
| EndToEndCloudTopicsTest  | test_write              |                                                                                                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68730#0197f4a9-71c8-494f-b601-8f8bf9b7ceaa | FAIL          | 0/1      |                                                                                                                                                            |
| CloudTopicsTest          | test_reconciler_uploads | {"cloud_storage_type": 2}                                                                                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68730#0197f4a9-71c9-4901-a622-dfdb41283c04 | FAIL          | 0/20     |                                                                                                                                                            |
| CloudTopicsTest          | test_reconciler_uploads | {"cloud_storage_type": 1}                                                                                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68730#0197f4a2-fbb1-4009-bc2b-1bfac023efd7 | FLAKY         | 1/21     | upstream reliability is '100.0'. current run reliability is '4.761904761904762'. drift is 95.2381 and the allowed drift is set to 50. The test should FAIL |
| CloudTopicsTest          | test_reconciler_uploads | {"cloud_storage_type": 1}                                                                                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68730#0197f4a9-71ca-4db2-961b-b5adeb990a41 | FAIL          | 0/18     |                                                                                                                                                            |
| RandomNodeOperationsTest | test_node_operations    | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68730#0197f4a2-fbbe-4328-b14a-c912dcb6824e | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS  |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68876">build#68876</a></summary>


| test_class                | test_method             | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:--------------------------|:------------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTest   | test_write              |                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68876#0197fbc5-ee19-48f7-8d39-5f029968f4b9 | FAIL          | 0/1      |                                                                                                                                                                        |
| EndToEndCloudTopicsTest   | test_write              |                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68876#0197fbc6-8b4a-4907-9d28-e1664a0e8e27 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| CloudTopicsTest           | test_reconciler_uploads | {"cloud_storage_type": 2} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68876#0197fbc5-ee1b-49cc-b47e-aed90d7f3522 | FAIL          | 0/19     |                                                                                                                                                                        |
| CloudTopicsTest           | test_reconciler_uploads | {"cloud_storage_type": 2} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68876#0197fbc6-8b4c-42db-aa80-0d65f17b69e3 | FLAKY         | 5/21     | upstream reliability is '100.0'. current run reliability is '23.809523809523807'. drift is 76.19048 and the allowed drift is set to 50. The test should FAIL           |
| CloudTopicsTest           | test_reconciler_uploads | {"cloud_storage_type": 1} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68876#0197fbc5-ee1b-494c-b80a-8977f48501e4 | FLAKY         | 1/19     |                                                                                                                                                                        |
| CloudTopicsTest           | test_reconciler_uploads | {"cloud_storage_type": 1} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68876#0197fbc6-8b43-41ff-b684-0355e39d6c99 | FLAKY         | 2/21     | upstream reliability is '100.0'. current run reliability is '9.523809523809524'. drift is 90.47619 and the allowed drift is set to 50. The test should FAIL            |
| CertificateRevocationTest | test_kafka              |                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68876#0197fbc5-ee1a-4478-919b-b2e5dc375fdc | FLAKY         | 20/21    | upstream reliability is '99.44903581267218'. current run reliability is '95.23809523809523'. drift is 4.21094 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68891">build#68891</a></summary>


| test_class              | test_method             | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:------------------------|:------------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTest | test_write              |                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68891#0197ff3a-0c3f-4751-8901-79ad3b5afb51 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| EndToEndCloudTopicsTest | test_write              |                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68891#0197ff3c-9305-4394-b818-e5349fec13e7 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| CloudTopicsTest         | test_reconciler_uploads | {"cloud_storage_type": 2} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68891#0197ff3a-0c40-4c66-bba8-52b4b5e2170c | FLAKY         | 2/21     | upstream reliability is '100.0'. current run reliability is '9.523809523809524'. drift is 90.47619 and the allowed drift is set to 50. The test should FAIL            |
| CloudTopicsTest         | test_reconciler_uploads | {"cloud_storage_type": 2} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68891#0197ff3c-9306-4586-9c01-4b00d6663edd | FLAKY         | 2/21     | upstream reliability is '100.0'. current run reliability is '9.523809523809524'. drift is 90.47619 and the allowed drift is set to 50. The test should FAIL            |
| CloudTopicsTest         | test_reconciler_uploads | {"cloud_storage_type": 1} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68891#0197ff3a-0c39-4518-9804-7138e5c46e19 | FLAKY         | 2/21     | upstream reliability is '100.0'. current run reliability is '9.523809523809524'. drift is 90.47619 and the allowed drift is set to 50. The test should FAIL            |
| CloudTopicsTest         | test_reconciler_uploads | {"cloud_storage_type": 1} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68891#0197ff3c-9307-4c46-aaec-9ef4bcf1bc2b | FLAKY         | 4/21     | upstream reliability is '100.0'. current run reliability is '19.047619047619047'. drift is 80.95238 and the allowed drift is set to 50. The test should FAIL           |
| ConsumerGroupTest       | test_group_lag_metrics  |                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68891#0197ff3a-0c3b-4ba9-a144-aeb1f3e26836 | FLAKY         | 20/21    | upstream reliability is '99.46666666666667'. current run reliability is '95.23809523809523'. drift is 4.22857 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68893">build#68893</a></summary>


| test_class                | test_method                                         | test_arguments                                                                                                                   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:--------------------------|:----------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageChunkReadTest | test_read_when_segment_size_smaller_than_chunk_size |                                                                                                                                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68893#019800cb-5299-4065-9079-73c81152bc5e | FLAKY         | 20/21    | upstream reliability is '97.93103448275862'. current run reliability is '95.23809523809523'. drift is 2.69294 and the allowed drift is set to 50. The test should PASS |
| DatalakeE2ETests          | test_json_schema_unicode                            | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "trino"}                                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68893#019800cb-5296-4cb2-8428-ef532ccb90be | FLAKY         | 20/21    | upstream reliability is '97.47899159663865'. current run reliability is '95.23809523809523'. drift is 2.2409 and the allowed drift is set to 50. The test should PASS  |
| TieredStorageTest         | test_tiered_storage                                 | {"cloud_storage_type_and_url_style": [1, "path"], "test_case": {"name": "(TS_Read == True, SpilloverManifestUploaded == True)"}} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68893#019800cb-5296-4cb2-8428-ef532ccb90be | FLAKY         | 20/21    | upstream reliability is '99.71671388101983'. current run reliability is '95.23809523809523'. drift is 4.47862 and the allowed drift is set to 50. The test should PASS |
| TieredStorageTest         | test_tiered_storage                                 | {"cloud_storage_type_and_url_style": [1, "path"], "test_case": {"name": "(TS_Read == True, TS_Timequery == True)"}}              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68893#019800cb-5298-47bb-b2d6-a01bb0366cb7 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69027">build#69027</a></summary>


| test_class      | test_method             | test_arguments     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                |
|:----------------|:------------------------|:-------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| MaintenanceTest | test_maintenance_sticky | {"use_rpk": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69027#01980f68-b221-4c8c-afa3-c48239d3c5df | FLAKY         | 20/21    | upstream reliability is '98.1859410430839'. current run reliability is '95.23809523809523'. drift is 2.94785 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69081">build#69081</a></summary>


| test_class             | test_method                    | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-----------------------|:-------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CompactionRecoveryTest | test_index_recovery            |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69081#01981267-a3aa-4c13-b306-b2bc868351a8 | FLAKY         | 20/21    | upstream reliability is '99.21875'. current run reliability is '95.23809523809523'. drift is 3.98065 and the allowed drift is set to 50. The test should PASS          |
| RaftAvailabilityTest   | test_controller_node_isolation |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69081#01981267-a3a8-499e-b4a5-a265e92b1c52 | FLAKY         | 20/21    | upstream reliability is '97.97752808988764'. current run reliability is '95.23809523809523'. drift is 2.73943 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69124">build#69124</a></summary>


| test_class      | test_method                           | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason   |
|:----------------|:--------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------|
| MasterTestSuite | test_chunk_dl_with_random_http_errors |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/69124#019814ac-da63-4a88-b56b-5cf04c8ab03b | FAIL          | 0/1      |          |
</details>
<!--/body-->

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/68684-->
<!--989b8f1b-58ce-4f22-878e-14fdbac21f30-->
<!--title-->
## Retry command for [Build#68684](https://buildkite.com/redpanda/redpanda/builds/68684)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cloud_topics/e2e_test.py::EndToEndCloudTopicsTest.test_write
```
<!--/body-->

rockwotj commented on src/v/cloud_topics/data_plane_impl.cc:
```diff
@@ -59,6 +59,8 @@ class impl
           std::make_unique<batch_cache>(&storage_api->local().log_mgr())) {}
 
     seastar::future<> start() override {
+        // Batcher
```
not batcher but batch cache?

rockwotj commented on src/v/cloud_topics/batch_cache/batch_cache.cc:
```diff
@@ -52,4 +68,20 @@ batch_cache::get(const model::ntp& ntp, model::offset o) {
     return std::nullopt;
 }
 
+void batch_cache::cleanup_index_entries() {
+    // NOTE: the memory is reclaimed asynchronously.  In some cases
+    // the index may no longer reference any live entries.  If this
+    // is the case we need to delete the batch_cache_index from the
+    // '_index'  collection to avoid accumulating orphaned entries.
+    chunked_vector<model::ntp> empty;
+    for (const auto& entry : _index) {
+        if (entry.second->empty()) {
+            empty.push_back(entry.first);
+        }
+    }
+    for (const auto& ntp : empty) {
+        _index.erase(ntp);
+    }
+}
```
Since the members here are chunked vectors and chunked_hash_map this could be large and yield stalls? We could move back to `absl::btree_map` and use the trick used for wasm GC: https://github.com/redpanda-data/redpanda/blob/1308cbd050d48f7ac4be24a835c52df42f3d8ebd/src/v/wasm/cache.cc#L36-L60

Lazin replied:
done

rockwotj commented on src/v/cloud_topics/batch_cache/batch_cache.cc:
```diff
@@ -16,9 +16,25 @@
 
 namespace experimental::cloud_topics {
 
+constexpr auto batch_cache_check_interval = std::chrono::seconds(10);
```
10 seconds is probably a little aggressive? Maybe 30 or 60 seconds? 🤷

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/68730-->
<!--e0770426-0b2e-4596-a4cd-178cc34756e6-->
<!--04c2f322-77d1-468c-974a-9d775c07b318-->
<!--f86148c2-6b38-41da-a0f1-3fbbbbc201ae-->
<!--title-->
## Retry command for [Build#68730](https://buildkite.com/redpanda/redpanda/builds/68730)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cloud_topics/e2e_test.py::EndToEndCloudTopicsTest.test_write
tests/rptest/tests/cloud_topics_test.py::CloudTopicsTest.test_reconciler_uploads@{"cloud_storage_type":1}
```
<!--/body-->

rockwotj commented:
Can we move the reader for L0 into the cloud_topics subsystem? Reconciliation is going to need it as well.

Lazin commented:
> Can we move the reader for L0 into the cloud_topics subsystem? Reconciliation is going to need it as well.

I'm not sure about this. In order to do this `cloud_topics` will have to have `cluster` as a dependency. Currently, stuff in `kafka/data` depends on `cloud_topics` and `cluster`. If `cloud_topics` module will also depend on `cluster` there will be a cyclic dependency. Also, the `reconciler` uses `cloud_topic_partition` now. It is the only component that depends on `cluster` because it's not really used in the `kafka/data` module. In other words, if the reader will stay where it is it will still be possible to use it in the reconciler.  Maybe it will be possible to migrate the reader, need to think more about it. But that could be done later.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/68876-->
<!--c2c674aa-5f24-4b61-8451-4161ef495e38-->
<!--58528b34-7dcb-4154-b609-2dc9e133da08-->
<!--4d5164dc-96e5-49e7-b051-c4888d88bd35-->
<!--0106abde-8e3c-473a-aa6b-3ff588bb78c3-->
<!--title-->
## Retry command for [Build#68876](https://buildkite.com/redpanda/redpanda/builds/68876)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cloud_topics/e2e_test.py::EndToEndCloudTopicsTest.test_write
tests/rptest/tests/cloud_topics_test.py::CloudTopicsTest.test_reconciler_uploads@{"cloud_storage_type":1}
tests/rptest/tests/cloud_topics_test.py::CloudTopicsTest.test_reconciler_uploads@{"cloud_storage_type":2}
```
<!--/body-->

Lazin committed:
ct: Represent timeouts using time_point

The existing code uses milliseconds which is not convenient and may lead
to overflow in some cases.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/68891-->
<!--58507726-192c-42f4-9ba2-1c8b0e19f19f-->
<!--af9d349d-481b-4ecf-a18e-71f939b634a1-->
<!--4aec553a-001f-4ebb-a4bb-704e62e5cd2f-->
<!--dd85205a-feb5-4acc-88b5-3a162b302211-->
<!--dab3d456-7b8c-44fe-85f4-1fe62fe227b0-->
<!--dd26f153-2b70-4d8f-b9de-7cc551190f71-->
<!--title-->
## Retry command for [Build#68891](https://buildkite.com/redpanda/redpanda/builds/68891)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cloud_topics_test.py::CloudTopicsTest.test_reconciler_uploads@{"cloud_storage_type":1}
tests/rptest/tests/cloud_topics_test.py::CloudTopicsTest.test_reconciler_uploads@{"cloud_storage_type":2}
tests/rptest/tests/cloud_topics/e2e_test.py::EndToEndCloudTopicsTest.test_write
```
<!--/body-->

Lazin commented:
I started another branch to try to extract relevant bits from the reader and the cloud_topic_partition into the src/v/cloud_topics. Called it 'frontend' for now, not sure if it makes sense to call it that way.

Lazin committed:
cluster: Add last_term to the kafka_result

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
raft: Add last_term to replicate_result

The field is used to communicate the term in which the batch was
replicated to the caller.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
cluster: Set last_term in rm_stm

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
raft: Propagate last_term in replicate_batcher

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
raft: Propagate last_term in replciate_entries_stm

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
cluster: Propagate last_term in cluster::partition

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Use last_term in cloud topics

Currently, the cloud_topic_partition caches term before replicating the
batch and assumes that the term didn't change. This is incorrect.
The fix is to use the last_term value propagated to the caller after the
replication.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

bharathv commented:
`last_term` commits lgtm.

rockwotj commented on src/v/kafka/data/cloud_topic_partition_reader.cc:
```diff
@@ -103,6 +103,19 @@ cloud_topic_partition_reader_impl::maybe_load_slices_from_cache() {
             // from S3.
             break;
         }
+        vlog(
+          kdlog.trace,
+          "Loaded batch from cache: {}",
+          batch.value().base_offset(),
+          batch.value().term());
+        if (batch.value().term() == model::term_id{-1}) {
+            // Skip bogus batch
+            vlog(
+              kdlog.warn,
+              "Batch with the invalid term is loaded: {}",
+              batch.value());
+            break;
+        }
```
When can this happen? Is this defensive - if so should it be a `vassert` or error log?

rockwotj reviewed (commented):
LGTM except for the bit where we load from cache and look for batches with term set to `-1`.

dotnwat commented on src/v/cloud_topics/batch_cache/batch_cache.cc:
```diff
@@ -0,0 +1,96 @@
+    return std::nullopt;
+}
+
+ss::future<> batch_cache::cleanup_index_entries() {
```
@rockwotj @Lazin do we have other signals here that we can use other than time to help keep the index size in check? for example if we were just considering l0 then after reconciliation we could remove (and evict) a range of batches (analgous to what we do when a segment is deleted).

it's not critical now, but my hunch is that this current strategy isn't going to scale and that this loop is going to show up in performance profiling.

dotnwat commented on src/v/cloud_topics/batch_cache/batch_cache.cc:
```diff
@@ -0,0 +1,96 @@
+        if (ok) {
+            it = new_it;
+        } else {
+            return;
```
more like `vassert(false)`

dotnwat reviewed (commented):
lgtm.

IIUC the primary use ase (at least right now) is to cache materialized batches for the L0 read path?
